src/components/MessageItem/index.tsx

```diff
@@ -0,0 +1,186 @@
+import MessageImage from './MessageImage';
+
+import { messageStyles } from './styles';
+
+const MessageItem: VFC<{
+  message: Amity.Message;
+  previousMessage: Amity.Message;
+  currentUserId: Amity.User['userId'];
+}> = ({ message: messageProp, previousMessage, currentUserId }) => {
+  const [user, setUser] = useState<Amity.User>();
+  const [avatar, setAvatar] = useState<Amity.File>();
+  const [messageData, setMessageData] = useState<Amity.Message>();
+
+  const {
+    colors,
+    fonts: {
+      regular: { fontFamily: regularFont },
+    },
+    dark,
+  } = useTheme();
+
+  const { messageId, userId, type } = messageProp;
+
+  useEffect(() => {
+    return observeUser(userId, ({ data }) => {
+      setUser(data);
+    });
+  }, [userId]);
+
+  useEffect(() => {
+    return observeMessage(messageId, ({ data }) => {
+      setMessageData(data);
+    });
+  }, [messageId]);
+
+  useEffect(() => {
+    if (user?.avatarFileId) {
+      return observeFile(user.avatarFileId, ({ data }) => {
+        setAvatar(data);
+      });
+    }
+
+    return undefined;
+  }, [user]);
+
+  const isSameDay = useMemo(
+    () => isSameDayUtil(messageProp, previousMessage),
+    [messageProp, previousMessage],
+  );
+  const isSameUser = useMemo(
+    () => isSameUserUtil(messageProp, previousMessage),
+    [messageProp, previousMessage],
+  );
+  const isSameThread = useMemo(() => isSameUser && isSameDay, [isSameDay, isSameUser]);
+
+  const MessageAvatar = useCallback(() => {
+    if (isSameThread) {
+      return null;
+    }
+
+    return avatar?.fileUrl ? (
+      <Avatar.Image source={{ uri: avatar?.fileUrl }} style={messageStyles.avatar} size={40} />
+    ) : (
+      <Avatar.Text
+        size={40}
+        style={messageStyles.avatar}
+        label={user?.displayName?.slice(0, 2).toUpperCase() ?? ''}
+      />
+    );
+  }, [avatar, isSameThread, user]);
+
+  const MessageHeader = useCallback(() => {
+    return isSameThread ? null : (
+      <View style={messageStyles.headerView}>
+        <Text
+          style={[messageStyles.standardFont, messageStyles.headerItem, messageStyles.username]}
+        >
+          {user?.displayName}
+        </Text>
+      </View>
+    );
+  }, [isSameThread, user]);
+
+  const Day = useCallback(() => {
+    return isSameDay ? null : (
+      <View style={messageStyles.timeContainerStyle}>
+        <Text style={messageStyles.timeTextStyle}>
+          {dayjs(messageProp.createdAt).format('YYYY MMMM DD')}
+        </Text>
+      </View>
+    );
+  }, [isSameDay, messageProp]);
+
+  const RenderMessageImage = () => {
+    if (type === 'image' && messageData) {
+      return <MessageImage message={messageData} />;
+    }
+    return null;
+  };
+
+  const RenderMessageText = () => {
+    if (type === 'text' && messageData) {
+      return (
+        <MessageText
+          message={messageData}
+          textStyle={{ color: colors.text, fontFamily: regularFont }}
+        />
+      );
+    }
+    return null;
+  };
```

Comment:
you may want to dig more into react works to understand why this is not correct.

you're doing the `return condition ? comp : null`  _inside_ the component instead of _outside_ the component. the different is that, _outside_ the component react will not create vdom node for no reason, but _inside_ the component, even if the component is null, there will be a vdom node + all the lifecycle calculations (on mount, etc etc)

Have a look here instead for the proper way to build this : https://github.com/AmityCo/Amity-Social-Cloud-Web-Sample-Apps/blob/main/create-react-app-simple-chat/src/components/MessageItem/index.js